tests/AnalyzeProperties.hs

```diff
@@ -0,0 +1,575 @@
+
+  ESimple TTime (Inj (IntAddTime x y)) ->
+    mkApp defAddTime [ESimple TTime x, ESimple TInt y]
+
+  ESimple TTime (Inj (DecAddTime x y)) ->
+    mkApp defAddTime [ESimple TTime x, ESimple TDecimal y]
+
+  ESimple TBool (Inj (Logical op args)) ->
+    mkApp (logicalOpToDef op) (ESimple TBool <$> args)
+
+  ESimple TInt     (CoreTerm (Lit x))
+    -> Just $ TLiteral (LInteger x) dummyInfo
+  ESimple TDecimal (CoreTerm (Lit x))
+    -> Just $ TLiteral (LDecimal (unMkDecimal x)) dummyInfo
+  ESimple TStr     (CoreTerm (Lit x))
+    -> Just $ TLiteral (LString (T.pack x)) dummyInfo
+  ESimple TBool    (CoreTerm (Lit x))
+    -> Just $ TLiteral (LBool x) dummyInfo
+  ESimple TTime    (CoreTerm (Lit x))
+    -> Just $ TLiteral (LTime (unMkTime x)) dummyInfo
+
+  ESimple TKeySet  (CoreTerm (Lit _x))
+    -> error "how do we translate keysets?"
+
+  tm -> error $ "TODO: toPact " ++ show tm
+
+  where
+    mkApp :: NativeDef -> [ETerm] -> Maybe (Pact.Term Pact.Ref)
+    mkApp (_, defTm) args = do
+      args' <- traverse toPact args
+      Just $ TApp (liftTerm defTm) args' dummyInfo
+
+    arithOpToDef :: ArithOp -> NativeDef
+    arithOpToDef = \case
+      Add -> addDef
+      Sub -> subDef
+      Mul -> mulDef
+      Div -> divDef
+      Pow -> powDef
+      Log -> logDef
+
+    unaryArithOpToDef :: UnaryArithOp -> NativeDef
+    unaryArithOpToDef = \case
+      Negate -> subDef
+      Sqrt   -> sqrtDef
+      Ln     -> lnDef
+      Exp    -> expDef
+      Abs    -> absDef
+      Signum -> undefined
+
+    roundingLikeOpToDef :: RoundingLikeOp -> NativeDef
+    roundingLikeOpToDef = \case
+      Round   -> roundDef
+      Ceiling -> ceilDef
+      Floor   -> floorDef
+
+    comparisonOpToDef :: ComparisonOp -> NativeDef
+    comparisonOpToDef = \case
+      Gt  -> gtDef
+      Lt  -> ltDef
+      Gte -> gteDef
+      Lte -> lteDef
+      Eq  -> eqDef
+      Neq -> neqDef
+
+    logicalOpToDef :: LogicalOp -> NativeDef
+    logicalOpToDef = \case
+      AndOp -> andDef
+      OrOp  -> orDef
+      NotOp -> notDef
+
+toPact' :: Applicative m => ETerm -> MaybeT m (Pact.Term Pact.Ref)
+toPact' = MaybeT . pure . toPact
+
+toAnalyze :: Pact.Type (Pact.Term Pact.Ref) -> Pact.Term Pact.Ref -> MaybeT IO ETerm
+toAnalyze ty tm = do
+  let cnst = TConst (Pact.Arg "tm" ty dummyInfo) "module" (Pact.CVRaw tm) Nothing dummyInfo
+      ref = Pact.Ref cnst
+  maybeConst <- lift $ Pact.runTC 0 False $ typecheckTopLevel ref
+  (_cTy, ast) <- case maybeConst of
+    (Pact.TopConst _info _name constTy constAst _meta, _tcState)
+      -> pure (constTy, constAst)
+    _ -> MaybeT $ pure Nothing
+  hoist generalize $
+    exceptToMaybeT $
+      fmap fst $
+        flip runStateT (TranslateState [] 0 0) $
+          runReaderT
+            (unTranslateM (translateNode ast))
+            (dummyInfo, mkTranslateEnv [])
+
+-- This is limited to simple types for now
+reverseTranslateType :: Type a -> Pact.Type b
+reverseTranslateType = \case
```

Comment:
this seems like the sort of thing that should live in `Pact.Analyze.Translate` / potentially be shared with the impl of stuff there